acyclic/src-3/acyclic/plugin/PluginPhase.scala

```diff
@@ -0,0 +1,205 @@
+package acyclic.plugin
+
+import acyclic.file
+import scala.collection.{SortedSet, mutable}
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.{CompilationUnit, report}
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Symbols.NoSymbol
+import dotty.tools.dotc.util.NoSource
+
+/**
+ * - Break dependency graph into strongly connected components
+ * - Turn acyclic packages into virtual "files" in the dependency graph, as
+ *   aggregates of all the files within them
+ * - Any strongly connected component which includes an acyclic.file or
+ *   acyclic.pkg is a failure
+ *   - Pick an arbitrary cycle and report it
+ * - Don't report more than one cycle per file/pkg, to avoid excessive spam
+ */
+class PluginPhase(
+  cycleReporter: Seq[(Value, SortedSet[Int])] => Unit,
+  force: => Boolean,
+  fatal: => Boolean
+)(using ctx: Context) extends GraphAnalysis[tpd.Tree] { t =>
+
+  private val pkgNameAccumulator = new tpd.TreeAccumulator[List[String]] {
+    @annotation.tailrec
+    private def definitivePackageDef(pkg: tpd.PackageDef): tpd.PackageDef =
+      pkg.stats.collectFirst { case p: tpd.PackageDef => p } match {
+        case Some(p) => definitivePackageDef(p)
+        case None => pkg
+      }
+
+    def apply(acc: List[String], tree: tpd.Tree)(using Context) = tree match {
+      case p: tpd.PackageDef => definitivePackageDef(p).pid.show :: acc
+      case _ => foldOver(acc, tree)
+    }
+  }
+
+  private def pkgName(unit: CompilationUnit) =
+    pkgNameAccumulator(Nil, unit.tpdTree).reverse.flatMap(_.split('.'))
+
+  private lazy val units = Option(ctx.run) match {
+    case Some(run) => run.units.toSeq.sortBy(_.source.content.mkString.hashCode())
+    case None => Seq()
+  }
+
+  private def hasImport(selector: String, tree: tpd.Tree): Boolean = {
+    val accumulator = new tpd.TreeAccumulator[Boolean] {
+      def apply(acc: Boolean, tree: tpd.Tree)(using Context): Boolean = tree match {
+        case tpd.Import(expr, List(sel)) =>
+          acc || (expr.symbol.toString == "object acyclic" && sel.name.show == selector)
+        case _ => foldOver(acc, tree)
+      }
+    }
+
+    accumulator(false, tree)
+  }
+
+  private val pkgObjectAccumulator = new tpd.TreeAccumulator[List[tpd.Tree]] {
+    def apply(acc: List[tpd.Tree], tree: tpd.Tree)(using Context): List[tpd.Tree] =
+      foldOver(
+        if (tree.symbol.isPackageObject) tree :: acc else acc,
+        tree
+      )
+  }
+
+  def findAcyclics() = {
```

Comment:
I wonder how much of the logic in this file can be shared between the Scala 2 and Scala 3 implementations? Like clearly the logic for working with symbols and types and trees is all bespoke, but it seems there's a lot of stuff working with `Value`s and `DepNode`s that can be shared if properly separated out from the compiler-specific logic

Reply:
Good call, I was able to share pretty much all of it by abstracting over `CompilationUnit`, `Tree`, and `Symbol`